stonesoup/dataassociator/tree.py

```diff
@@ -0,0 +1,176 @@
+                indexes = tree.query_ball_point(
+                    meas_pred.state_vector.ravel(),
+                    r=self.max_distance)
+            else:
+                _, indexes = tree.query(
+                    meas_pred.state_vector.ravel(),
+                    k=self.number_of_neighbours,
+                    distance_upper_bound=self.max_distance)
+
+            for index in np.atleast_1d(indexes):
+                if index != len(detections_list):
+                    track_detections[track].add(detections_list[index])
+
+        return {track: self.hypothesiser.hypothesise(
+            track, track_detections[track], time, **kwargs)
+            for track in tracks}
+
+
+class TPRTreeMixIn(DataAssociator):
+    measurement_model = Property(MeasurementModel)
+    horizon_time = Property(datetime.timedelta)
+    vel_mapping = Property(np.ndarray, default=None)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if self.vel_mapping is None:
+            self.pos_mapping = self.measurement_model.mapping[::2]
+            self.vel_mapping = self.measurement_model.mapping[1::2]
+        else:
+            self.pos_mapping = self.measurement_model.mapping
+
+        # Create tree
+        tree_property = rtree.index.Property(
+            type=rtree.index.RT_TPRTree,
+            tpr_horizon=self.horizon_time.total_seconds(),
+            dimension=len(self.pos_mapping))
+        self._tree = rtree.index.RtreeContainer(properties=tree_property)
+        self._coords = dict()
+
+    def _track_tree_coordinates(self, track):
+        state_vector = track.state_vector[self.pos_mapping, :]
+        state_delta = 3 * np.sqrt(
+            np.diag(track.covar)[self.pos_mapping].reshape(-1, 1))
+        vel_vector = track.state_vector[self.vel_mapping, :]
+        vel_delta = 3 * np.sqrt(
+            np.diag(track.covar)[self.vel_mapping].reshape(-1, 1))
+
+        min_pos = (state_vector - state_delta).ravel()
+        max_pos = (state_vector + state_delta).ravel()
+        min_vel = (vel_vector - vel_delta).ravel()
+        max_vel = (vel_vector + vel_delta).ravel()
+
+        return ((*min_pos, *max_pos), (*min_vel, *max_vel),
+                track.timestamp.timestamp())
+
+    def generate_hypotheses(self, tracks, detections, time, **kwargs):
+        # No need for tree here.
+        if not tracks:
+            return dict()
+
+        c_time = None
+        for track in sorted(tracks.union(self._tree),
+                            key=attrgetter('timestamp')):
+            if c_time is None:
+                c_time = track.timestamp
+            if track not in self._tree:
+                self._coords[track] = self._track_tree_coordinates(track)
+                self._tree.insert(track, self._coords[track])
+            elif track not in tracks:
+                coords = self._coords[track][:-1] \
+                    + ((self._coords[track][-1], c_time.timestamp()),)
```

Comment:
Note `timestamp()` will be in local time, and cause issues with changing time zones. Should be converted to UTC first.